core/src/main/scala/io/finch/Decode.scala

```diff
@@ -3,77 +3,94 @@ package io.finch
 import java.util.UUID
 
 import com.twitter.util.{Return, Throw, Try}
-import shapeless.{::, Generic, HNil}
+import shapeless.{::, Generic, HNil, Witness}
 
 /**
  * An abstraction that is responsible for decoding the value of type `A`.
  */
-trait Decode[A] {
-  def apply(s: String): Try[A]
+trait Decode[I, A] {
+  type ContentType <: String
+  def apply(i: I): Try[A]
 }
 
 trait LowPriorityDecodeInstances {
 
+  type Aux[I, A, CT <: String] = Decode[I, A] { type ContentType = CT }
+
+  type ApplicationJson[I, A] = Aux[I, A, Witness.`"application/json"`.T]
+  type TextPlain[I, A] = Aux[I, A, Witness.`"text/plain"`.T]
```

Comment:
@ImLiar if you force the type to string here it will compile (at least compiles my code):
```
type TextPlain[A] = Aux[String,A, Witness.`"text/plain"`.T]
```
i have tried with implicit view but it did not work: 
```
type TextPlain[I <: String, A] = Aux[I, A, Witness.`"text/plain"`.T]
```
we can do the same for applicationJson. first leave it as string and then make it c.t.u.Buf when it is ready.
Again there might be better.